app/vmctl/remoteread/remoteread.go

```diff
@@ -0,0 +1,265 @@
+package remoteread
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+const (
+	defaultReadTimeout = 30 * time.Second
+	remoteReadPath     = "/api/v1/read"
+	healthPath         = "/-/healthy"
+)
+
+// StreamCallback is a callback function for processing time series
+type StreamCallback func(series prompb.TimeSeries) error
+
+// Client is an HTTP client for reading
+// time series via remote read protocol.
+type Client struct {
+	addr     string
+	c        *http.Client
+	user     string
+	password string
+}
+
+// Config is config for remote read.
+type Config struct {
+	// Addr of remote storage
+	Addr string
+	// ReadTimeout defines timeout for HTTP write request
+	// to remote storage
+	ReadTimeout time.Duration
+	// Username is the remote read username, optional.
+	Username string
+	// Password is the remote read password, optional.
+	Password string
+}
+
+// Filter is used for request remote read data by filter
+type Filter struct {
+	StartTimestampMs int64
+	EndTimestampMs   int64
+	Label            string
+	LabelValue       string
+}
+
+// NewClient returns client for
+// reading time series via remote read protocol.
+func NewClient(cfg Config) (*Client, error) {
+	if cfg.Addr == "" {
+		return nil, fmt.Errorf("config.Addr can't be empty")
+	}
+	if cfg.ReadTimeout == 0 {
+		cfg.ReadTimeout = defaultReadTimeout
+	}
+
+	c := &Client{
+		c: &http.Client{
+			Timeout:   cfg.ReadTimeout,
+			Transport: http.DefaultTransport.(*http.Transport).Clone(),
+		},
+		addr:     strings.TrimSuffix(cfg.Addr, "/"),
+		user:     cfg.Username,
+		password: cfg.Password,
+	}
+	return c, nil
+}
+
+// Read fetch data from remote read source
+func (c *Client) Read(ctx context.Context, filter *Filter, streamCb StreamCallback) error {
+	query, err := c.query(filter)
+	if err != nil {
+		return fmt.Errorf("error prepare stream query: %w", err)
+	}
+	req := &prompb.ReadRequest{
+		Queries: []*prompb.Query{
+			query,
+		},
+		AcceptedResponseTypes: []prompb.ReadRequest_ResponseType{prompb.ReadRequest_STREAMED_XOR_CHUNKS},
+	}
+	data, err := proto.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("unable to marshal read request: %w", err)
+	}
+
+	const attempts = 5
+	b := snappy.Encode(nil, data)
+	for i := 0; i < attempts; i++ {
```

Comment:
I'm not sure, that retry is needed. Because you will try to fetch all stream from the beginning. I think, it must be handled by remote stream library.

Reply:
removed attemts